File: Terremesh.Converter/trc/Terremesh/QuadricErrorMetric/QuadricErrorMetricMethod.cpp
#include "QuadricErrorMetricMethod.h"

#include "../Math/Matrix.h"
#include "../Math/Vec3.h"

namespace Terremesh
{
namespace QuadricErrorMetric
{
	void QuadricErrorMetricMethod::Process(Remesh::Mesh& mesh, double targetRatio, IProgressListener* listener)
	{
		int totalTriangles = mesh.GetTriangles().size();
		int trianglesLimit = (int)(targetRatio * totalTriangles);

		Process(mesh, trianglesLimit, listener);
	}

	void QuadricErrorMetricMethod::Process(Remesh::Mesh& mesh, int targetTriangles, IProgressListener* listener)
	{
		m_Vertices = mesh.GetVertices();
		m_Triangles = mesh.GetTriangles();

		Initialize(listener);
		Remesh(targetTriangles, listener);

		mesh.SetVertices(m_Vertices);
		mesh.SetTriangles(m_Triangles);
	}

	void QuadricErrorMetricMethod::Initialize(IProgressListener* listener)
	{
		if (listener != nullptr)
		{
			listener->OnStarted("Initialize quadrics");
		}

		// Initialize error metrics for each vertex available.
		for (auto i = 1; i <= (int)m_Vertices.size(); ++i)
		{
			// By empty error metric.
			m_ErrorMetrics.insert(
				std::make_pair(
					(int)i,
					ErrorMetric()
				)
			);
		}

		// For each triangle compute plane metric and add it to neighbor vertex
		for (auto it = m_Triangles.begin(); it != m_Triangles.end(); ++it)
		{
			for (auto vertex = 0; vertex < 3; ++vertex)
			{
				auto& vertexMetric = m_ErrorMetrics[it->Vertices[vertex]];
				auto& planeMetric = ErrorMetric(it->Plane);

				// Adding error metrics.
				ErrorMetric::Add(vertexMetric, vertexMetric, planeMetric);
			}
		}

		if (listener != nullptr)
		{
			listener->OnCompleted("Initialize quadrics");
		}
	}

	void QuadricErrorMetricMethod::SelectValidPairs(double treshold, IProgressListener* listener)
	{
		if (listener != nullptr)
		{
			listener->OnStarted("Selecting pairs");
		}

		// For each triangle
		for (auto it = m_Triangles.begin(); it != m_Triangles.end(); ++it)
		{
			// Compute edge costs (edge 01)
			VertexPair pair;
			pair.first = std::min(it->Vertices[0], it->Vertices[1]);
			pair.second = std::max(it->Vertices[0], it->Vertices[1]);

			if (m_Edges.find(pair) == m_Edges.end())
			{
				m_Edges.insert(std::make_pair(pair, ComputeError(pair)));
			}

			// Compute edge costs (edge 12)
			pair.first = std::min(it->Vertices[1], it->Vertices[2]);
			pair.second = std::max(it->Vertices[1], it->Vertices[2]);

			if (m_Edges.find(pair) == m_Edges.end())
			{
				m_Edges.insert(std::make_pair(pair, ComputeError(pair)));
			}

			// Compute edge costs (edge 20)
			pair.first = std::min(it->Vertices[2], it->Vertices[0]);
			pair.second = std::max(it->Vertices[2], it->Vertices[0]);

			if (m_Edges.find(pair) == m_Edges.end())
			{
				m_Edges.insert(std::make_pair(pair, ComputeError(pair)));
			}
		}

		if (listener != nullptr)
		{
			listener->OnCompleted("Selecting pairs");
		}

		// Check if virtual pairs are enabled 
		/// @bug Not working as expected.
		if (m_EnableVirtualPairs)
		{
			if (listener != nullptr)
			{
				listener->OnStarted("Selecting virtual pairs");
			}

			// Search for vertex pairs with distance lesser than treshold
			for (auto i = 1; i < (int)m_Vertices.size(); ++i)
			{
				for (auto j = i + 1; j < (int)m_Vertices.size(); ++j)
				{
					if (Math::Vec3::Distance(m_Vertices[i].Position, m_Vertices[j].Position) < treshold)
					{
						VertexPair pair(i, j);
						m_Edges.insert(std::make_pair(pair, ComputeError(pair)));
					}
				}
			}

			if (listener != nullptr)
			{
				listener->OnCompleted("Selecting virtual pairs");
			}
		}
	}

	double QuadricErrorMetricMethod::ComputeError(Remesh::VertexId id1, Remesh::VertexId id2, Math::Vec3& error)
	{
		ErrorMetric edge;

		Math::Vec3 vertex;

		// Get metrics for involved vertices
		auto e1 = m_ErrorMetrics[id1];
		auto e2 = m_ErrorMetrics[id2];

		// Add and assume they represent edge error metric
		ErrorMetric::Add(edge, e1, e2);

		// Log when they aren't symmetric
		if (!edge.GetMatrix().IsSymetric())
		{
			std::cerr << "The result matrix is not symmetric" << std::endl;
		}

		// Compute delta edge
		ErrorMetric delta = edge;
		{
			Math::Matrix m = delta.GetMatrix();
			{
				m.M41 = 0.0;
				m.M42 = 0.0;
				m.M43 = 0.0;
				m.M44 = 1.0;
			}
			delta.SetMatrix(m);
		}

		double minError = 0.0;

		// If matrix is not invertible
		if (std::abs(delta.GetMatrix().Determinant()) <= 1e-5)
		{
			// Take two vertices and center between them
			Math::Vec3 v1 = m_Vertices[id1].Position;
			Math::Vec3 v2 = m_Vertices[id2].Position;
			Math::Vec3 v3;
			Math::Vec3::Center(v3, v1, v2);
			
			// And evaluate costs
			double e1 = edge.Evaluate(v1);
			double e2 = edge.Evaluate(v2);
			double e3 = edge.Evaluate(v3);

			// And choose wisely
			minError = std::min(std::min(e1, e2), e3);

			if (minError == e1)
			{
				vertex = v1;
			}
			else if (minError == e2)
			{
				vertex = v2;
			}
			else if (minError == e3)
			{
				vertex = v3;
			}
		}
		else
		{
			// Otherwise compute vertex from matrix.
			vertex = delta.GetMatrix().GetVector();
		}

		error = vertex;
		minError = edge.Evaluate(vertex);
		return minError;
	}

	void QuadricErrorMetricMethod::Remesh(int targetTriangles, IProgressListener* listener)
	{
		SelectValidPairs(0.1, listener);

		if (listener != nullptr)
		{
			listener->OnStarted("Remesh");
		}

		// Compute total and remaining triangles count.
		int totalTriangles = m_Triangles.size();
		int remainingTriangles = m_Triangles.size() - targetTriangles;


		Math::Vec3 error;

		// Until we don't reached remaining triangles count.
		while ((int)m_Triangles.size() > remainingTriangles)
		{
			if (listener != nullptr)
			{
				listener->OnStep(
					targetTriangles - (m_Triangles.size() - remainingTriangles),
					targetTriangles);
			}
			
			double minError = (double)std::numeric_limits<int>::max();

			EdgeErrorContainer::iterator itMinError;

			// Find cheapest edge
			for (auto it = m_Edges.begin(); it != m_Edges.end(); ++it)
			{
				if (it->second < minError)
				{
					minError = it->second;
					itMinError = it;
				}
			}

			// Compute error for pair
			VertexPair pairMinError = itMinError->first;

			ComputeError(pairMinError, error);

			m_Vertices[pairMinError.first].Position = error;
			
			// Compute error metric
			ErrorMetric::Add(
				m_ErrorMetrics[pairMinError.first],
				m_ErrorMetrics[pairMinError.first],
				m_ErrorMetrics[pairMinError.second]);

			// And for each triangle
			for (auto it = m_Triangles.begin(); it != m_Triangles.end();)
			{
				// And for each vertex in triangle
				for (auto j = 0; j < 3; ++j)
				{
					if (it->Vertices[j] == pairMinError.second)
					{
						if (
							(it->Vertices[0] == pairMinError.first) ||
							(it->Vertices[1] == pairMinError.first) ||
							(it->Vertices[2] == pairMinError.first))
						{
							// Erase
							it = m_Triangles.erase(it);
						}
						else
						{
							// Or not
							it->Vertices[j] = pairMinError.first;
							++it;
						}
						break;
					}
					else if (j == 2)
					{
						++it;
					}
				}
			}

			// And erase second vertex - it's merged now with first
			m_Vertices.erase(pairMinError.second);

			// Update involved edges costs - set as 0
			for (auto it = m_Edges.begin(); it != m_Edges.end(); /*++it*/)
			{
				auto pair = it->first;

				if ((pair.first == pairMinError.second) && (pair.second != pairMinError.first))
				{
					it = m_Edges.erase(it);

					m_Edges.insert(std::make_pair(
						VertexPair(
							std::min(pairMinError.first, pair.second),
							std::max(pairMinError.first, pair.second)
						),
						0.0
					));
				}
				else if ((pair.second == pairMinError.second) && (pair.first != pairMinError.first))
				{
					it = m_Edges.erase(it);

					m_Edges.insert(std::make_pair(
						VertexPair(
							std::min(pairMinError.first, pair.first),
							std::max(pairMinError.first, pair.first)
						),
						0.0
					));
				}
				else
				{
					++it;
				}
			}

			m_Edges.erase(itMinError);

			// Recompute all involved edges costs.
			for (auto it = m_Edges.begin(); it != m_Edges.end(); ++it)
			{
				auto pair = it->first;

				if (pair.first == pairMinError.first)
				{
					it->second = ComputeError(pairMinError.first, pair.second);
				}

				if (pair.second == pairMinError.first)
				{
					it->second = ComputeError(pairMinError.first, pair.first);
				}
			}
		}

		if (listener != nullptr)
		{
			listener->OnCompleted("Remesh");
		}
	}
}
}

File: Terremesh.Mesh/Quadric/QuadricErrorMetricMethod.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;
using Terremesh.Mesh.XnaFramework;

namespace Terremesh.Mesh.Quadric
{
    public class QuadricErrorMetricMethod : IRemeshingMethod
    {
        public void Process(ref Mesh mesh, double targetRatio, IProgressListener listener = null)
        {
            if ((targetRatio <= 0.0) || (targetRatio > 1.0))
            {
                throw new ArgumentOutOfRangeException("targetRatio");
            }

            int targetTriangles = (int)(mesh.Faces.Count * targetRatio);

            Process(ref mesh, targetTriangles, listener);
        }

        public void Process(ref Mesh mesh, int targetTriangles, IProgressListener listener = null)
        {
            if ((targetTriangles <= 0) || (targetTriangles > mesh.Faces.Count))
            {
                throw new ArgumentOutOfRangeException("targetTriangles");
            }

            Clear();

            m_Vertices = mesh.Vertices;
            m_Faces = mesh.Faces;

            ComputeInitialQuadrics();
            Contract(targetTriangles, listener);

            mesh.Vertices = m_Vertices;
            mesh.Faces = m_Faces;
        }

        private void Clear()
        {
            m_Vertices = null;
            m_Faces = null;
            m_Errors.Clear();
            m_Quadrics.Clear();
            m_Splits.Clear();
        }

        private void Contract(int targetTriangles, IProgressListener listener = null)
        {
            SelectValidPairs();

            if (listener != null)
            {
                listener.OnStarted("Compacting");
            }

            int totalTriangles = m_Faces.Count - targetTriangles;
            int currentTriangle = 0;
            while (m_Faces.Count > targetTriangles)
            {
                if (listener != null)
                {
                    listener.OnStep(currentTriangle, totalTriangles);
                }
                ++currentTriangle;
                double minError = (double)int.MaxValue;

                //KeyValuePair<VertexPair, double> min;
                VertexPair minPair = new VertexPair() { First = 0, Second = 0 };

                foreach (KeyValuePair<VertexPair, double> e in m_Errors)
                {
                    if (e.Value < minError)
                    {
                        minError = e.Value;
                        minPair = e.Key;
                    }
                }

                Vector3 error;
                ComputeError(minPair.First, minPair.Second, out error);

#if false
                VertexSplit split = new VertexSplit();

                //var fv

                split.First.Position = m_Vertices[minPair.First].Position;
                split.Second.Position = m_Vertices[minPair.Second].Position;
                split.Target.Position = error;
                m_Splits.Add(split);
#endif
                var vertex = m_Vertices[minPair.First];
                vertex.Position = error;

                m_Quadrics[minPair.First] = m_Quadrics[minPair.First] + m_Quadrics[minPair.Second];

                for (int i = m_Faces.Count - 1; i != 0; )
                {
                    var face = m_Faces[i];

                    for (int j = 0; j < 3; ++j)
                    {
                        if (face[j] == minPair.Second)
                        {
                            if (face[0] == minPair.First || face[1] == minPair.First || face[2] == minPair.First)
                            {
                                m_Faces.Remove(face);
                            }
                            else
                            {
                                face[j] = minPair.First;
                            }
                            --i;
                            break;
                        }
                        else if (j == 2)
                        {
                            --i;
                        }
                    }
                }

                m_Vertices.Remove(minPair.Second);

                KeyValuePair<VertexPair, double> pair;

#if false
                for (int iter = m_Errors.Count - 1; iter != 0; )
                {
                    pair = m_Errors.ElementAt(iter);

                    if (pair.Key.First == minPair.Second && pair.Key.Second != minPair.First)
                    {
                        m_Errors.Remove(m_Errors.ElementAt(iter).Key);

                        m_Errors.Add(
                            new VertexPair() { First = Math.Min(minPair.First, pair.Key.Second), Second = Math.Max(minPair.First, pair.Key.Second) },
                            0.0);
                        --iter;
                    }
                    else if (pair.Key.Second == minPair.Second && pair.Key.First != minPair.First)
                    {
                        m_Errors.Remove(m_Errors.ElementAt(iter).Key);

                        m_Errors.Add(
                            new VertexPair() { First = Math.Min(minPair.First, pair.Key.First), Second = Math.Max(minPair.First, pair.Key.First) },
                            0.0);
                        --iter;
                    }
                    else
                    {
                        --iter;
                    }
                }
#else
                for (int it = 0; it < m_Errors.Count; ++it)
                {
                    pair = m_Errors.ElementAt(it);

                    if (pair.Key.First == minPair.Second && pair.Key.Second != minPair.First)
                    {
                        m_Errors.Remove(m_Errors.ElementAt(it).Key);

                        var key = new VertexPair()
                        {
                            First = Math.Min(minPair.First, pair.Key.Second),
                            Second = Math.Max(minPair.First, pair.Key.Second)
                        };

                        if (!m_Errors.ContainsKey(key))
                        {
                            m_Errors.Add(
                                key,
                                0.0);
                        }
                    }
                    else if (pair.Key.Second == minPair.Second && pair.Key.First != minPair.First)
                    {
                        m_Errors.Remove(m_Errors.ElementAt(it).Key);

                        var key = new VertexPair()
                        {
                            First = Math.Min(minPair.First, pair.Key.First),
                            Second = Math.Max(minPair.First, pair.Key.First)
                        };

                        if (!m_Errors.ContainsKey(key))
                        {
                            m_Errors.Add(
                                key,
                                0.0);
                        }
                    }
                }
#endif
                m_Errors.Remove(minPair);

                for (int it = 0; it < m_Errors.Count; ++it)
                {
                    var key = m_Errors.ElementAt(it).Key;

                    if (key.First == minPair.First)
                    {
                        m_Errors[key] = ComputeError(minPair.First, key.Second);
                    }

                    if (key.Second == minPair.First)
                    {
                        m_Errors[key] = ComputeError(minPair.First, key.First);
                    }
                }
                /*foreach (var e in m_Errors)
                {
                    var p = e.Key;
                    if (p.First == minPair.First)
                    {
                        m_Errors[p] = ComputeError(minPair.First, p.Second);
                    }

                    if (p.Second == minPair.First)
                    {
                        m_Errors[p] = ComputeError(minPair.First, p.First);
                    }
                }*/

            }
            if (listener != null)
            {
                listener.OnComplete("Compacting");
            }
        }

        private void ComputeInitialQuadrics()
        {
            for (int i = 0; i < m_Vertices.Count; ++i)
            {
                m_Quadrics.Add(i + 1, ErrorMetric.Empty);
            }

            for (int i = 0; i < m_Faces.Count; ++i)
            {
                for (int j = 0; j < 3; ++j)
                {
                    var face = m_Faces.ElementAt(i);
                    var index = face[j];

#if DEBUG
                    if (!m_Quadrics.ContainsKey(index))
                    {
                        System.Diagnostics.Debug.WriteLine("Failed {0}", index);
                    }
#endif
                    m_Quadrics[index] += new ErrorMetric(face.Plane);
                }
            }
        }

        private void SelectValidPairs(double level = 0.12)
        {
            for (int i = 0; i < m_Faces.Count; ++i)
            {
                var face = m_Faces.ElementAt(i);

                VertexPair pair;
                pair.First = Math.Min(face[0], face[1]);
                pair.Second = Math.Max(face[0], face[1]);

                if (!m_Errors.ContainsKey(pair))
                {
                    m_Errors.Add(pair, ComputeError(pair.First, pair.Second));
                }

                pair.First = Math.Min(face[0], face[2]);
                pair.Second = Math.Max(face[0], face[2]);

                if (!m_Errors.ContainsKey(pair))
                {
                    m_Errors.Add(pair, ComputeError(pair.First, pair.Second));
                }

                pair.First = Math.Min(face[1], face[2]);
                pair.Second = Math.Max(face[1], face[2]);

                if (!m_Errors.ContainsKey(pair))
                {
                    m_Errors.Add(pair, ComputeError(pair.First, pair.Second));
                }
            }

            if (m_EnableVirtualPairs)
            {
                for (int i = 1; i < m_Vertices.Count; ++i)
                {
                    for (int j = i + 1; j < m_Vertices.Count; ++j)
                    {
                        if (Vector3.Distance(m_Vertices[i].Position, m_Vertices[j].Position) < level)
                        {
                            VertexPair pair;
                            pair.First = i;
                            pair.Second = j;
                            if (!m_Errors.ContainsKey(pair))
                            {
                                m_Errors.Add(pair, ComputeError(i, j));
                            }
                        }
                    }
                }
            }
        }

        private double ComputeError(int id1, int id2)
        {
            Vector3 unused;
            return ComputeError(id1, id2, out unused);
        }

        private double ComputeError(int id1, int id2, out Vector3 error)
        {
            ErrorMetric edge = ErrorMetric.Empty;
            ErrorMetric delta = ErrorMetric.Empty;

            Vector3 vertex = new Vector3();

            edge = m_Quadrics[id1] + m_Quadrics[id2];

            delta = edge;
            Matrix m = delta.Matrix;
            {
                m.M41 = 0.0f;
                m.M42 = 0.0f;
                m.M43 = 0.0f;
                m.M44 = 1.0f;
            }
            delta.Matrix = m;

            double minError = 0.0;

            if (Math.Abs(delta.Matrix.Determinant()) <= 1e-5)
            {
                Vector3 v1 = m_Vertices[id1].Position;
                Vector3 v2 = m_Vertices[id2].Position;
                Vector3 v3 = Vector3Extension.Mean(v1, v2);

                double e1 = edge.Evaluate(v1);
                double e2 = edge.Evaluate(v2);
                double e3 = edge.Evaluate(v3);

                minError = Math.Min(Math.Min(e1, e2), e3);

                if (minError == e1)
                {
                    vertex = v1;
                }
                else if (minError == e2)
                {
                    vertex = v2;
                }
                else if (minError == e3)
                {
                    vertex = v3;
                }
            }
            else
            {
                double det = (double)delta.Matrix.IndexedDeterminant(0, 1, 2, 4, 5, 6, 8, 9, 10);
                vertex.X = (float)(-1.0 / det * (delta.Matrix.IndexedDeterminant(1, 2, 3, 5, 6, 7, 9, 10, 11)));
                vertex.Y = (float)(1.0 / det * (delta.Matrix.IndexedDeterminant(0, 2, 3, 4, 6, 7, 8, 10, 11)));
                vertex.Z = (float)(-1.0 / det * (delta.Matrix.IndexedDeterminant(0, 1, 3, 4, 5, 7, 8, 9, 11)));
            }

            error = vertex;
            minError = edge.Evaluate(vertex);
            return minError;
        }

        public bool EnableVirtualPairs
        {
            get { return m_EnableVirtualPairs; }
            set { m_EnableVirtualPairs = value; }
        }

        private Dictionary<int, Vertex> m_Vertices = null;
        private List<Triangle> m_Faces = null;
        private Dictionary<int, ErrorMetric> m_Quadrics = new Dictionary<int, ErrorMetric>();
        private Dictionary<VertexPair, double> m_Errors = new Dictionary<VertexPair, double>();
        private List<VertexSplit> m_Splits = new List<VertexSplit>();
        private bool m_EnableVirtualPairs = false;
    }
}
